table/partitioned_filter_block.cc

```diff
@@ -9,76 +9,61 @@
+    : FullFilterBlockBuilder(prefix_extractor, whole_key_filtering,
+                             filter_bits_builder),
+      index_on_filter_block_builder_(index_block_restart_interval),
+      part_index_builder_(part_index_builder) {}
+
+PartitionedFilterBlockBuilder::~PartitionedFilterBlockBuilder() {}
 
-void PartitionIndexBuilder::AddIndexEntry(
-    std::string* last_key_in_current_block,
-    const Slice* first_key_in_next_block, const BlockHandle& block_handle) {
-  sub_index_builder_->AddIndexEntry(last_key_in_current_block,
-                                    first_key_in_next_block, block_handle);
-  num_indexes++;
-  if (UNLIKELY(first_key_in_next_block == nullptr)) {  // no more keys
-    entries_.push_back({std::string(*last_key_in_current_block),
-                        std::unique_ptr<IndexBuilder>(sub_index_builder_)});
-    sub_index_builder_ = nullptr;
-  } else if (num_indexes % index_per_partition_ == 0) {
-    entries_.push_back({std::string(*last_key_in_current_block),
-                        std::unique_ptr<IndexBuilder>(sub_index_builder_)});
-    sub_index_builder_ =
-        CreateIndexBuilder(sub_type_, comparator_, prefix_extractor_,
-                           index_block_restart_interval_, index_per_partition_);
+// FullFilterBlockBuilder
+void PartitionedFilterBlockBuilder::MayBeCutAFilterBlock() {
+  if (!part_index_builder_->ShouldCutFilterBlock()) {
+    return;
   }
+  filter_gc.push_back(std::unique_ptr<const char[]>(nullptr));
+  Slice filter = filter_bits_builder_->Finish(&filter_gc.back());
+  std::string& index_key = part_index_builder_->GetPartitionKey();
+  filters.push_back({index_key, filter});
+}
+
+void PartitionedFilterBlockBuilder::AddKey(const Slice& key) {
+  MayBeCutAFilterBlock();
+  filter_bits_builder_->AddKey(key);
 }
 
-Status PartitionIndexBuilder::Finish(
-    IndexBlocks* index_blocks, const BlockHandle& last_partition_block_handle) {
-  assert(!entries_.empty());
-  // It must be set to null after last key is added
-  assert(sub_index_builder_ == nullptr);
-  if (finishing == true) {
-    Entry& last_entry = entries_.front();
+Slice PartitionedFilterBlockBuilder::Finish(
+    const BlockHandle& last_partition_block_handle, Status* status) {
+  if (finishing_filters == true) {
+    FilterEntry& last_entry = filters.front();
     std::string handle_encoding;
     last_partition_block_handle.EncodeTo(&handle_encoding);
-    index_block_builder_.Add(last_entry.key, handle_encoding);
-    entries_.pop_front();
+    // This is the full key vs. the user key in the filters. We assume that
+    // user key <= full key
+    index_on_filter_block_builder_.Add(last_entry.key, handle_encoding);
+    filters.pop_front();
   }
-  // If there is no sub_index left, then return the 2nd level index.
-  if (UNLIKELY(entries_.empty())) {
-    index_blocks->index_block_contents = index_block_builder_.Finish();
-    return Status::OK();
+  MayBeCutAFilterBlock();
```

Comment:
when `finishing_filters=true` then MayBeCutAFitlerBlock will be essentially noop. but It would good to make it clear in the code.